README.md
# monitor_printf
## Library for printf to the Arduino IDE serial monitor window or to other ports

This library makes it easy to use printf()-style function calls in your Arduino C++ code to print messages to the Arduino IDE serial monitor window or to other serial ports.

## Quick start

Here is sample code if you want to quickly get going:

```
#include <monitor_printf.h>

// Set this to 1 to enable debug output to the serial monitor.
// Set this to 0 when releasing code to use without the IDE, USB port, and serial monitor.
#define USE_MONITOR_PORT 1

// Define the port to be used by the global instance named "the_serial_monitor".
#if USE_MONITOR_PORT
#define MONITOR_PORT &Serial  // Enable printf output to the serial monitor port identified by variable "Serial".
#else
#define MONITOR_PORT NULL     // Disable printf output when using WITHOUT the IDE, USB port, and serial monitor.
#endif

void setup() {
  monitor.begin(MONITOR_PORT, 115200);
  monitor.printf("MyProgram is running\n");
}
```

## Basics

The library consists of a .cpp/.h file pair that define a C++ class named `monitor_printf` whose member functions are used to initialize a serial port for printf output and to perform printfs to the port.

### Include file

To use the library, include the library header file in your code file:

```
  #include <monitor_printf.h>
```

This #include creates one global instance of the `monitor_printf` class, named **`the_monitor_printf`**. It

It also #defines the symbol **`monitor`** to equal `the_monitor_printf`. Either `the_monitor_printf` or `monitor` can be used to call `monitor_printf` member functions to do printfs to the serial monitor. For example:

```
  monitor.printf("Hello world!\n");  // ** preferred **

  the_monitor_printf.printf("Hello world!\n"); // ** optional **
```

Since `monitor` might possibly already be used elsewhere in your code, a method is provided to not define it. Simply define the symbol `DONT_DEFINE_monitor` before including the library header file, and then optionally #define a different short symbol name to use to do printfs:

```
// Skip defining 'monitor'.
#define DONT_DEFINE_monitor
#include <monitor_printf.h>

// Define something shorter than 'the_monitor_printf' to use, so we can do mon.printf(...);
#define mon the_monitor_printf
```

### Choosing a serial port OR no output

After the #include, it is recommended that you #define symbol USE_MONITOR_PORT equal to 1 when running the code with the IDE and serial monitor window, and to 0 when running it with no IDE, no USB port, and no serial monitor. After that #define, you should add a #if USE_MONITOR_PORT, and when it is 1, #define MONITOR_PORT to the address of the HardwareSerial variable for the serial port to be used (usually **`&Serial`**), then after a #else, #define it as NULL, with a final #endif. For example:

```
// Set this to 1 to enable debug output to the serial monitor.
// Set this to 0 when releasing code to use without the IDE, USB port, and serial monitor.
#define USE_MONITOR_PORT 1

// Define the port to be used by the global instance named "the_serial_monitor".
#if USE_MONITOR_PORT
#define MONITOR_PORT &Serial  // Enable printf output to the serial monitor port identified by variable "Serial".
#else
#define MONITOR_PORT NULL     // Disable printf output when using WITHOUT the IDE, USB port, and serial monitor.
#endif
```

The value **`MONITOR_PORT`** is used in the call to the monitor_printf `begin()` member function to properly initialize the object for running with or without the serial monitor.

**It is necessary to change the #define USE_MONITOR_PORT each time you switch between running with the Arduino IDE and serial monitor (defined as 1), versus without it (defined as 0).** The above example shows USE_MONITOR_PORT set to 1 for running with the IDE and serial monitor. When you change it to 0, this causes printf() calls to do nothing. You don't need to remove the printf calls from your code.

### Call member function `begin()` to initialize

The `monitor_printf` member function **`begin()`** initializes the class instance for printing, including initialization of the HardwareSerial object that communicates with the serial monitor window. Alternatively, `begin()` can disable the `printf()` function, used when running code without the USB port and IDE monitor window.

The `begin()` function MUST be called to initialize before printing, usually as the first thing in the **`setup()`** function. It has several arguments, all optional:

```
void begin(HardwareSerial* serial = NULL, unsigned long baud = 115200, byte config = SERIAL_8N1);
```

The first argument, **`serial`**, is normally set to **`MONITOR_PORT`**, the #define described above. This causes `serial` to be **`&Serial`** when running with the USB port connected to the Arduino IDE, or to be **`NULL`** when running without the serial port and IDE.

The second and third arguments, **`baud`** and **`config`**, are only used by `begin()` when `MONITOR_PORT` is *not* NULL. In that case, `begin()` calls the serial port's own `begin()` function to initialize serial communications, using `baud` and `config` to set serial speed and data/parity/stop bits. You should set `baud` to the serial port speed you want to use, or omit the argument to use the default setting of 115,200 bps speed. Typically `config` is omitted to use its default of `SERIAL_8N1`.

If `MONITOR_PORT` is `NULL`, this *disables the printf() function, so when it is called it does nothing*. When you are running your code outside the Arduino IDE, with no USB port and no serial monitor, it is **crucial that `MONITOR_PORT` (and therefore the `serial` argument) be set to `NULL`**. Otherwise, your code will hang when it tries to initialize a non-existent serial port. This is why it is recommended that you define `USE_MONITOR_PORT` at the top of the file where it is easily visible to help remind you to change it whenever you switch between running with the IDE and without it, followed by a define of `MONITOR_PORT` to the appropriate value depending on how `USE_MONITOR_PORT` is set.

In summary, the call to `begin()` will typically look like this if you want to run the serial port at 9600 bps:

```
void setup() {
  monitor.begin(MONITOR_PORT, 9600);
}
```

### Call member function `printf()` to print

The `monitor_printf` member function **`printf()`** prints a character string to the serial monitor window, using a standard printf-style format string to control the formatting of the text, and with optional arguments that contain values to be printed or are arguments used by format string "% specifiers".

The ``printf()`` member function uses an internal C++ library function (``vsnprintf()``) to format your text string for output. It uses the ``format`` argument you specify, plus any additional arguments that follow it as needed to process the printf() "%-specifiers".

Here is an example that includes both the initialization call to `begin()` and two calls to `printf()`:

```
void setup() {
  monitor.begin(MONITOR_PORT, 9600);
  monitor.printf("MyProgram is running\n");
}

void loop() {

int x = 123;
long Z = 123456789;
char c = 'A';
char* p = "Goodbye";

monitor.printf("Value of x: %d   Z: %ld   c: %c   p: %s\n", x, Z, c, p);
delay(1000);
```

The first call to `printf()` inside the `setup()` function shows a simple printf call to print a string followed by a newline character to start a new line on the serial monitor.

The second call inside the `loop()` function is a more complex printf call that uses **%-specifiers** to print integer (%d), long integer (%ld), character (%c), and string (%s) values. The actual variables containing those values are specified as additional arguments following the format string, in the same order as they appear in the "%-specifiers".

If you don't have a good working knowledge of "%-specifiers", including features such as field width and justification, it is recommended that you look them up and study them a bit. As an example, the specifier "%-25s" prints a string left-justified (the '-' sign) in a field 25 characters wide. As another example, arguments `("%*d", K, i)` prints integer in a field whose width is given by integer K (the '*' says to look for an argument to specify the field width).

Although most standard printf() functionality should be present, I've noticed that **%u** is missing (use **`%d`** instead, not ideal but it often works for most cases).

Also, something often of great importance that is typically missing is the ability to format and print ``float`` and ``double`` values. These are apparently disabled in standard versions of Arduino libraries, but can be enabled through some mechanism that I have not yet figured out. If you need to print floating point values, you may want to use my **``floatToString``** library.

## Additional details for those interested

### Using more instances of the serial_monitor class for other serial ports

The above descriptions implied that the serial_monitor class specifically uses the `Serial` object, which is the standard serial port object used to communicate with the IDE's serial monitor window. However, the `monitor_printf` class actually allows any serial port to be used, so long as the port's variable is an instance of class `HardwareSerial` (or a class derived from it). The `HardwareSerial` class is defined in the standard Arduino library to encompass the functionality needed for a serial port, and the standard `Serial` object is an instance of the `HardwareSerial` class.

If you are using a microcomputer with hardware that supports multiple serial ports, there is a good chance that the Arduino run-time code defines a variable for each one, all derived from `HardwareSerial`. Often these variables are named `Serial1`, `Serial2`, etc. You can try putting one of those variables in your code and checking to see if the code compiles to find out if it is there or not. Right-clicking on a such a system variable within the IDE and choosing "Go to Definition" opens the file where the variable is defined, and this is often useful to see how it is defined and what class it is.

The `serial` argument to the `begin()` member function can be something other than `MONITOR_PORT`. To use a `monitor_printf` instance for doing printfs to some other serial port, you define a new instance and then call its `begin()` function, passing it the address of that port. For example, to use both the normal `monitor` variable with the serial port variable `Serial` and to use port `Serial2` with a second `monitor_printf` instance named `ser2`:

```
#include <monitor_printf.h>

// Set this to 1 to enable debug output to the serial monitor.
// Set this to 0 when releasing code to use without the IDE, USB port, and serial monitor.
#define USE_MONITOR_PORT 1

// Define the port to be used by the global instance named "the_serial_monitor".
#if USE_MONITOR_PORT
#define MONITOR_PORT &Serial  // Enable printf output to the serial monitor port identified by variable "Serial".
#else
#define MONITOR_PORT NULL     // Disable printf output when using WITHOUT the IDE, USB port, and serial monitor.
#endif

// A monitor_printf instance for using port Serial2.
monitor_printf ser2;

void setup() {
  monitor.begin(MONITOR_PORT, 115200);
  monitor.printf("MyProgram is running\n");

  ser2.begin(&Serial2, 9600);
}

void loop() {
  ...
  // Print string S to Serial2 with value i.
  ser2.printf("%s %d", S, i);
  ...
}
```

You might also want to use another serial port but **not use the serial port connected to the serial monitor window.** To NOT define the global instance variable 'the_monitor_printf', #define DONT_DEFINE_the_monitor_printf before including the library header file. You should ALSO #define DONT_DEFINE_monitor. For example:

```
// Skip defining 'monitor'.
#define DONT_DEFINE_monitor
// Skip defining 'the_monitor_printf'.
#define DONT_DEFINE_the_monitor_printf
#include <monitor_printf.h>

// A monitor_printf instance for using port Serial2.
monitor_printf ser2;

void setup() {
  ser2.begin(&Serial2, 9600);
}

void loop() {
  ...
  // Print string S to Serial2 with value i.
  ser2.printf("%s %d", S, i);
  ...
}
```

The `monitor_printf` class C++ **constructor** function has one optional argument: a pointer to the `HardwareSerial` instance to use with the port. If the argument is omitted, NULL is used and then the serial port is specified when `begin()` is called. You can use the constructor argument to specify the serial port to be used with the instance, INSTEAD OF specifying it in the `begin()` call. In that case, there are no arguments  to the `begin()` call (`serial` defaults to NULL and other arguments are not used). This is almost the same as the previous example, *except the `begin()` function skips port initialization. You must initialize the serial port yourself.* For example:

```
// Skip defining 'monitor'.
#define DONT_DEFINE_monitor
// Skip defining 'the_monitor_printf'.
#define DONT_DEFINE_the_monitor_printf
#include <monitor_printf.h>

// A monitor_printf instance for using port Serial2.
monitor_printf ser2(&Serial2);

void setup() {
  // Initialize Serial2 port.
  Serial2.begin(9600);
  // Initialize ser2 for printf to Serial2.
  ser2.begin();
}

void loop() {
  ...
  // Print string S to Serial2 with value i.
  ser2.printf("%s %d", S, i);
  ...
}
```

### Buffer size grows automatically or is changed with `setBufSize()`

The ``printf()`` member function formats the string to be output into a buffer, then writes the string to the output using the `Serial.write()` function. An initial buffer is automatically allocated, and if that initial buffer is not big enough for the character string you are printing, the printf() member function automatically reallocates the buffer to a larger size that will hold it. It currently never reduces the buffer size, but if for some reason you wanted to do that, you can do it like this:

```
// Output a very long string.
monitor.printf("very long string....");
// Reallocate buffer to a small one. It will grow as needed.
monitor.setBufSize(20);
```

This uses the ``setBufSize()`` member function, which frees the existing buffer, if any, then allocates a new buffer
of the specified size. It is used by ``begin()`` and ``printf()`` when buffer allocation is needed.

## Contact

If you find a problem, please submit an issue report [here](https://github.com/tedtoal/monitor_printf/issues/new/choose).

You can contact me (the author) at [ted@tedtoal.net](ted@tedtoal.net)


CHANGELOG.md
# monitor_printf change log
---

## Table of Contents

* [Release 1.6.1](#release-161)
* [Release 1.5.1](#release-151)
* [Release v1.4.1](#release-v141)
* [Release v1.3.1](#release-v131)
* [Release v1.2.1](#release-v121)
* [Release v1.1.1](#release-v111)
---

### Release 1.6.1

1. Again no changes. Same as with 1.5.1, trying to get library manager to pick
    up the release.

### Release 1.5.1

1. No changes. The only reason for doing this release is that the Arduino
    Library Manager did not pick up release V1.4.1, giving the comment "Release
    monitor_printf:1.4.1 already loaded, skipping". I suspect the cause may be
    that my release tag values were of the form "V1.4.1" instead of "1.4.1", so
    I am trying it again, changing the release tag.

### Release v1.4.1

1. Bug fixed in automatic buffer reallocation.

2. The 'enable' argument in the begin() function is removed and replaced with
    new argument 'serial'. See README.md for recommended way of supplying this
    argument value (using #defines USE_MONITOR_PRINTF and MONITOR_PRINTF) to
    best cope with switching from use of IDE/USB to not. **** BE SURE TO CHANGE
    YOUR CALLS TO begin() TO USE THE NEW FIRST ARG NAMED 'serial' INSTEAD OF
    'enable' ****

3. Add support for DONT_DEFINE_the_monitor_printf, see README.md.

4. README.md has been extensively updated.

### Release v1.3.1

1. Not documented in release V1.2.1 was that the begin() function buf_size
    argument was removed. **** BE SURE TO CHANGE YOUR CALLS TO begin() TO REMOVE
    THE buf_size ARGUMENT ****

### Release v1.2.1

1. Add README, CHANGELOG, and LICENSE files.
2. Add HardwareSerial* argument to constructor to allow support for any serial port.
3. Change buffer allocation to automatically resize buffer and start with a default size.

### Release v1.1.1

1. Initial release.



src/monitor_printf.cpp
/*
  monitor_printf.cpp - Defines class monitor_printf containing a function to do
  printfs to the serial monitor, with one global instance of it.
  Created by Ted Toal, July 5, 2023.
  Released into the public domain.


  Software License Agreement (BSD License)

  Copyright (c) 2023 Ted Toal
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
  3. Neither the name of the copyright holders nor the
  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ''AS IS'' AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include <Arduino.h>
#include <monitor_printf.h>
#include <stdarg.h>
#include <stdio.h>

/**************************************************************************/

#ifndef DONT_DEFINE_the_monitor_printf
monitor_printf the_monitor_printf;
#endif

/**************************************************************************/

void monitor_printf::begin(HardwareSerial* serial, unsigned long baud,
    byte config) {
  if (serial != NULL)
    _serial = serial;
  setEnabled(_serial != NULL);
  setBufSize(INITIAL_MONITOR_PRINTF_BUF_SIZE);
  // Only initialize serial port if 'serial' argument is not NULL.
  if (serial != NULL) {
    // This delay seems to be necessary, at least for SAMD21 hardware.
    delay(1000);
    _serial->begin(baud, config);
    // Discard anything in serial receive buffer.
    while (! *_serial)
      ;
    // Final delay.
    delay(200);
  }
}

/**************************************************************************/

void monitor_printf::setBufSize(uint16_t buf_size) {
  if (_buf != NULL) {
    free(_buf);
    _buf = NULL;
    _buf_size = 0;
  }
  if (buf_size > 0) {
    _buf_size = buf_size;
    _buf = (char *)malloc(_buf_size);
  }
}

/**************************************************************************/
void monitor_printf::printf(const char* format, ...) {
  if (_enabled) {
    va_list args;
    va_start(args, format);
    int N = vsnprintf(_buf, _buf_size, format, args);
    // Reallocate a larger buffer if it was too small.
    if (N >= _buf_size) {
      setBufSize(N + 1);
      N = vsnprintf(_buf, _buf_size, format, args);
    }
    va_end(args);
    if (N >= 0)
      _serial->write(_buf);
    else
      _serial->write("vsnprintf() error occurred!\n");
  }
}

// -------------------------------------------------------------------------


src/monitor_printf.h
/*
  monitor_printf.h - Defines class monitor_printf containing a function to do
  printfs to a serial port, with one global instance that prints to the Arduino
  IDE serial monitor port.
  Created by Ted Toal, July 5, 2023.
  Released into the public domain.


  Software License Agreement (BSD License)

  Copyright (c) 2023 Ted Toal
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
  3. Neither the name of the copyright holders nor the
  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ''AS IS'' AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#ifndef monitor_printf_h
#define monitor_printf_h

#include <Arduino.h>

#define INITIAL_MONITOR_PRINTF_BUF_SIZE 150

class monitor_printf {

protected:
  HardwareSerial* _serial;
  bool _enabled;
  uint16_t _buf_size;
  char *_buf;

public:
  /**************************************************************************/
  /*!
    @brief    Constructor.
    @param    serial  Pointer to HardwareSerial object for the serial port to be
                      printf-ed to, or NULL to defer setting the port until
                      begin() is called.
    @note     Example:
                #include <monitor_printf.h>
                // Set this to 1 to enable debug output to the serial monitor.
                // Set this to 0 when releasing code to use without the IDE, USB port, and serial monitor.
                #define USE_MONITOR_PORT 1
                // Define the port to be used by the global instance named "the_serial_monitor".
                #if USE_MONITOR_PORT
                #define MONITOR_PORT &Serial  // Enable printf output to the serial monitor port identified by variable "Serial".
                #else
                #define MONITOR_PORT NULL     // Disable printf output when using WITHOUT the IDE, USB port, and serial monitor.
                #endif

                void setup() {
                  monitor.begin(MONITOR_PORT, 9600);
                  monitor.printf("Printing to monitor, can you see it?\n");
                }
                void loop() {
                  monitor.printf("Numbers from 1 to 10:\n");
                  for (uint8_t i = 0; i < 10; i++)
                    monitor.printf(" %d", i);
                  monitor.printf("\n");
                  delay(10000);
                }
  */
  /**************************************************************************/
  monitor_printf(HardwareSerial* serial = NULL) : _serial(serial),
    _enabled(false), _buf_size(0), _buf(NULL) {}

  /**************************************************************************/
  /*!
    @brief      Destructor. Release buffer memory.
  */
  /**************************************************************************/
  ~monitor_printf() { setBufSize(0); }

  /**************************************************************************/
  /*!
    @brief    Initialize monitor_printf object.
    @param    serial      Pointer to HardwareSerial object for the serial port
                          to be printf-ed to, or NULL to keep the port that was
                          set with the constructor 'serial' argument.
    @param    baud        serial port speed in bits/second.
    @param    config      SERIAL_ constant defining data, parity, and stop bits.
    @note     When 'serial' is not NULL, this initializes the HardwareSerial
              object with baud and config, empties the Serial input buffer, does
              short delays to ensure smooth operation, and enables printfs. If
              'serial' is NULL, that initialization is not done and printfs
              remain disabled.
    @note     Use NULL for your standard monitor serial port when running
              WITHOUT A USB connection (in your final system with no USB), else
              your system will hang because the "Serial" object doesn't exist in
              that case.
  */
  /**************************************************************************/
  void begin(HardwareSerial* serial = NULL, unsigned long baud = 115200,
    byte config = SERIAL_8N1);

  /**************************************************************************/
  /*!
    @brief    Get current enable state.
    @returns  true if enabled, false if disabled.
  */
  /**************************************************************************/
  const bool isEnabled(void) { return (_enabled); }

  /**************************************************************************/
  /*!
    @brief    Set enable state.
    @param    enable    true to enable printing, false makes printf do nothing.
  */
  /**************************************************************************/
  void setEnabled(bool enable) { _enabled = (_serial != NULL) && enable; }

  /**************************************************************************/
  /*!
    @brief    Get current buffer size.
    @returns  Buffer size in chars including 1 for final nul, 0 if no buffer.
  */
  /**************************************************************************/
  const uint16_t getBufSize(void) { return (_buf_size); }

  /**************************************************************************/
  /*!
    @brief    Set buffer size.
    @param    buf_size  Buffer size in characters including 1 for final nul, or
                        0 to only deallocate current buffer, if any.
    @note     It is not necessary for the user to call this function, as the
              buffer size is managed automatically, see printf() note.
  */
  /**************************************************************************/
  void setBufSize(uint16_t buf_size);

  /**************************************************************************/
  /*!
    @brief    print string to serial monitor using printf-style format string.
    @param    format    Format string just like for printf(). Note that some
                        specifiers may not be supported under some versions and
                        settings of some hardware platforms.
    @param    ...       The values to be printed, each corresponding to format
                        string % specifiers or their arguments.
    @note     If the buffer size (buf_size arguments above) is not large enough
              to hold the string resulting from the printf formatting, it is
              automatically reallocated here to a large enough size.
  */
  /**************************************************************************/
  void printf(const char* format, ...);
};

/**************************************************************************/
/*!
  @brief    An instance of a monitor_printf class object is defined with a
            "serial" serial port pointer equal to NULL.
  @note     #define DONT_DEFINE_the_serial_monitor before #including file to not
            create the 'the_monitor_printf' instance.
  @note     You would normally set the serial port to "Serial" when calling the
            begin() function, unless you are releasing a version without a USB
            serial port, in which case you would set it to NULL.
  @note     You can also refer to the_monitor_printf using the #define 'monitor'
            below (unless that is disabled with DONT_DEFINE_monitor).
*/
/**************************************************************************/
#ifndef DONT_DEFINE_the_monitor_printf
extern monitor_printf the_monitor_printf;
#endif

// #define DONT_DEFINE_monitor before #including file to not define 'monitor'
#ifndef DONT_DEFINE_monitor
#define monitor the_monitor_printf
#endif

#endif // monitor_printf_h
